### PR TITLE
[modes] make python arg parser also check for python key

### LINKED ISSUE #90652: Assertion error triggered by CrossRefFakeMode
### 🐛 Describe the bug

Running with `TORCH_COMPILE_DEBUG=1`, if the model contains ops without meta tensor support, I got the failure due to assertion from
https://github.com/pytorch/pytorch/blob/7ec1cb855373dbc011b403adb87d943e395d8d2a/torch/_subclasses/meta_utils.py#L203-L205

The issue can be demonstrated with a simple test case below:
```python
import torch
from torch.utils._mode_utils import no_dispatch

with torch._subclasses.FakeTensorMode():
    a = torch.rand([100])
    with torch._subclasses.CrossRefFakeMode():
        # below logic from run_fallback_kernel
        with no_dispatch():
            b = torch.zeros_like(a) # <- exception thrown here
```
The chain looks something like `FakeTensorMode` -> `run_fallback_kernel` -> `no_dispatch()` -> `torch.zeros_like` -> `CrossRefFakeMode` -> `from_tensor` -> ... -> assertion error expecting `no_dispatch` off.

The error is triggered from `CrossRefFakeMode` which will be turned on if the `TORCH_COMPILE_DEBUG=1` is on.

My question:
1. What's the contract of calling functions on tensor subclasses with __torch_dispatch__ with `no_dispatch`? Is it allowed to call `torch.zeros_like(a)` under `no_dispatch` when `a` is a tensor subclass like proxy tensor?
2. What's the purpose of `CrossRefFakseMode`? Is it expecting real tensors or can also support fake tensors?

### Versions

Collecting environment information...
PyTorch version: 2.0.0a0+git184f6b5
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: Ubuntu 20.04.5 LTS (x86_64)
GCC version: (Ubuntu 9.4.0-1ubuntu1~20.04.1) 9.4.0
Clang version: Could not collect
CMake version: version 3.25.0
Libc version: glibc-2.31

Python version: 3.8.13 (default, Mar 28 2022, 11:38:47)  [GCC 7.5.0] (64-bit runtime)
Python platform: Linux-5.11.0-27-generic-x86_64-with-glibc2.17
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] bert-pytorch==0.0.1a4
[pip3] clip-anytorch==2.5.0
[pip3] CoCa-pytorch==0.0.6
[pip3] dalle2-pytorch==1.10.5
[pip3] ema-pytorch==0.0.10
[pip3] functorch==1.14.0a0+408bcf1
[pip3] intel-extension-for-pytorch==1.13.0+cpu
[pip3] mypy-extensions==0.4.3
[pip3] numpy==1.23.5
[pip3] pytorch-transformers==1.2.0
[pip3] pytorch-warmup==0.1.1
[pip3] rotary-embedding-torch==0.1.5
[pip3] torch==2.0.0a0+git7552e12
[pip3] torch-fidelity==0.3.0
[pip3] torch-struct==0.5
[pip3] torchaudio==0.14.0a0+41c4fd6
[pip3] torchmetrics==0.10.3
[pip3] torchrec-nightly==2022.11.21
[pip3] torchtext==0.15.0a0+58dad70
[pip3] torchvision==0.15.0a0+3100734
[pip3] torchx==0.3.0
[pip3] vector-quantize-pytorch==0.10.12
[conda] bert-pytorch              0.0.1a4                   dev_0    <develop>
[conda] clip-anytorch             2.5.0                    pypi_0    pypi
[conda] coca-pytorch              0.0.6                    pypi_0    pypi
[conda] dalle2-pytorch            1.10.5                   pypi_0    pypi
[conda] ema-pytorch               0.0.10                   pypi_0    pypi
[conda] functorch                 1.14.0a0+408bcf1          pypi_0    pypi
[conda] intel-extension-for-pytorch 1.13.0+cpu                dev_0    <develop>
[conda] mkl                       2022.1.0           hc2b9512_224  
[conda] mkl-include               2022.2.1                 pypi_0    pypi
[conda] mkl-static                2022.2.1                 pypi_0    pypi
[conda] numpy                     1.23.5                   pypi_0    pypi
[conda] pytorch-transformers      1.2.0                    pypi_0    pypi
[conda] pytorch-warmup            0.1.1                    pypi_0    pypi
[conda] rotary-embedding-torch    0.1.5                    pypi_0    pypi
[conda] torch                     2.0.0a0+git7552e12           dev_0    <develop>
[conda] torch-fidelity            0.3.0                    pypi_0    pypi
[conda] torch-struct              0.5                      pypi_0    pypi
[conda] torchaudio                0.14.0a0+41c4fd6          pypi_0    pypi
[conda] torchmetrics              0.10.3                   pypi_0    pypi
[conda] torchrec-nightly          2022.11.21               pypi_0    pypi
[conda] torchtext                 0.15.0a0+58dad70          pypi_0    pypi
[conda] torchvision               0.15.0a0+3100734          pypi_0    pypi
[conda] torchx                    0.3.0                    pypi_0    pypi
[conda] vector-quantize-pytorch   0.10.12                  pypi_0    pypi

### PR DESCRIPTION
**Pull request by samdow**
Fixes #90652

Previously, we had assumed that the only way to call `handle_torch_function_no_python_arg_parser` was through the Python key. This is no longer true with FakeTensor. Specifically `_like` functions will call `.device()` on FakeTensors when the args list is being parsed. In order to respect that the mode stack shouldn't run when the python key is off, this just adds that a check that the python key is on/the torch_function equivalent to that function

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #91573
* #90770

### CODE REVIEW

**Review by ezyang — APPROVED** [review]

**Comment by ezyang** [comment]
onnx failures are spurious so if everything else is good you can force merge
